### PR TITLE
noms log: add ability to log any path, not just dataset

### PR DESCRIPTION
**Pull request by aboodman**

### CODE REVIEW

**Comment by aboodman** [comment]
@willhite only review the last commit -- the others are being reviewed separately.

**Comment by kalman** [comment]
rebase? (I haven't looked at any of the changes, you can squash everything together now and push -f if you want)

File: cmd/noms/noms_log.go [hunk 1]
```diff
@@ -11,11 +11,14 @@ import (
 	"math"
 	"strings"
 
+	"os"
```
Comment (kalman) [inline cmd/noms/noms_log.go]: import with other imports

File: cmd/noms/noms_log.go [hunk 2]
```diff
@@ -61,18 +64,27 @@ func setupLogFlags() *flag.FlagSet {
...
-		d.CheckErrorNoUsage(err)
+	resolved := cfg.ResolvePathSpec(args[0])
+	sp, err := spec.ForPath(resolved)
+	d.CheckErrorNoUsage(err)
+	pinned, ok := sp.Pin()
+	if !ok {
+		fmt.Fprintf(os.Stderr, "Cannot resolve spec: %s\n", args[0])
+		return 1
 	}
+	database := pinned.GetDatabase()
```
Comment (kalman) [inline cmd/noms/noms_log.go]: `defer pinned.Close()` (not `defer database.Close()`)
In fact we should also have a `defer sp.Close()` too.

File: cmd/noms/noms_log.go [hunk 3]
```diff
@@ -61,18 +64,27 @@ func setupLogFlags() *flag.FlagSet {
...
 	defer database.Close()
 
-	if value == nil {
-		d.CheckErrorNoUsage(fmt.Errorf("Object not found: %s", args[0]))
+	absPath := pinned.Path
+	d.Chk.NotEmpty(absPath.Hash) // because we pinned it
```
Comment (kalman) [inline cmd/noms/noms_log.go]: better to be in the habit of doing `d.PanicIfTrue` even though in this case the interface boxing isn't a big deal. But this assertion doesn't seem so useful.

File: cmd/noms/noms_log.go [hunk 4]
```diff
@@ -298,13 +315,27 @@ func writeDiffLines(node LogNode, db datas.Database, maxLines, lineno int, w io.
...
 	parentCommit := parent.(types.Ref).TargetValue(db).(types.Struct)
-	err = diff.PrintDiff(pw, parentCommit.Get(datas.ValueField), node.commit.Get(datas.ValueField), true)
-	mlw.MaxLines = 0
-	if err != nil {
-		d.PanicIfNotType(err, writers.MaxLinesErr)
-		pw.NeedsPrefix = true
-		pw.Write([]byte("...\n"))
-		err = nil
+	old := path.Resolve(parentCommit)
+	neu := path.Resolve(node.commit)
```
Comment (kalman) [inline cmd/noms/noms_log.go]: You could consider resolving these in parallel.

**Comment by aboodman** [comment]
ptal

File: cmd/noms/noms_log.go [hunk 5]
```diff
@@ -9,9 +9,10 @@ import (
 	"fmt"
 	"io"
 	"math"
+	"os"
 	"strings"
 
-	"os"
+	"sync"
```
Comment (kalman) [inline cmd/noms/noms_log.go]: up with the other imports...

File: cmd/noms/noms_log.go [hunk 6]
```diff
@@ -315,8 +316,16 @@ func writeDiffLines(node LogNode, path types.Path, db datas.Database, maxLines,
...
+	var old, neu types.Value
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	resolve := func(commit types.Value, dest *types.Value) {
+		*dest = path.Resolve(commit)
+		wg.Done()
+	}
+	go resolve(parentCommit, &old)
+	go resolve(node.commit, &neu)
+	wg.Wait()
```
Comment (kalman) [inline cmd/noms/noms_log.go]: More fun would be:
```go
var old, neu types.Value
functions.All(
  func() { old = path.Resolve(parentCommit) },
  func() { neu = path.Resolve(node.commit) }
)
```

File: cmd/noms/noms_log.go (hunk 5 above)
Comment (aboodman) [inline cmd/noms/noms_log.go]: dang VS keeps doing that. Fixed.

File: cmd/noms/noms_log.go (hunk 6 above)
Comment (aboodman) [inline cmd/noms/noms_log.go]: Ah, nice.

**Comment by aboodman** [comment]
OK mom.

**Pull request merged**